_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```
@@ -85,6 +87,141 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
+            m_frameId(frameId), m_frameAllocator(frameAllocator)
+        {
+            if (m_frameAllocator == nullptr)
+            {
+                throw vp9_exception(UMC::UMC_ERR_FAILED);
+            }
+            Lock();
+        }
+
+        FrameLocker& operator=(FrameLocker &&locker)
+        {
+            Unlock();
+            m_frameId = locker.m_frameId;
+            m_frameAllocator = locker.m_frameAllocator;
+            try
+            {
+                Lock();
+            }
+            catch (const vp9_exception &ex)
+            {
+                m_frameId = -1;
+                throw;
+            }
+            return *this;
+        }
+
+        ~FrameLocker()
+        {
+            try
+            {
+                Unlock();
+            }
+            catch(...)
+            {
+                (void)MFX_STS_TRACE(MFX_ERR_UNKNOWN);
+            }
+        }
+    };
+
+    std::map<UMC::FrameMemID, FrameLocker[8]> m_lockedRefFrames;
+    std::vector<UMC::FrameMemID> m_decodedFramesIDs;
+    std::shared_ptr<mfx_UMC_FrameAllocator> m_frameAllocator;
+
+    void FreeReferenceFrames(UMC::FrameMemID currFrame)
+    {
+        size_t erasedCount = m_lockedRefFrames.erase(currFrame);
+
+        if (erasedCount == 0)
+        {
+            throw vp9_exception(UMC::UMC_ERR_FAILED);
+        }
+    }
+public:
+    ReferenceFrameStorage(std::shared_ptr<mfx_UMC_FrameAllocator> &frameAllocator):m_frameAllocator(frameAllocator)
+    {
+    }
+
+    void LockReferenceFrames(UMC::FrameMemID currFrame, const UMC::FrameMemID ref_frame_map[UMC_VP9_DECODER::NUM_REF_FRAMES])
+    {
+        try
+        {
+            // FreeReferenceFrames must return error, currFrame should not be in m_lockedRefFrames map
+            // but if it is there free resources (decrease the references).
+            FreeReferenceFrames(currFrame);
+            (void)MFX_STS_TRACE(MFX_ERR_UNKNOWN);
+        }
+        catch (const vp9_exception &ex)
+        {
```

exception is just ignored?

---

Fixed.